python/context.cpp

```diff
@@ -16,67 +18,158 @@
 
 namespace pyarb {
 
+namespace {
+auto is_nonneg_int = [](auto&& t){ return ((t==int(t) and t>=0)); };
+}
+
+// A Python shim that holds the information that describes an arb::proc_allocation.
+struct proc_allocation_shim {
+    using opt_int = arb::util::optional<int>;
+
+    opt_int gpu_id = {};
+    int num_threads = 1;
+
+    proc_allocation_shim(): proc_allocation_shim(1, pybind11::none()) {}
+
+    proc_allocation_shim(int threads, pybind11::object gpu) {
+        set_num_threads(threads);
+        set_gpu_id(gpu);
+    }
+
+    // getter and setter (in order to assert when being set)
+    void set_gpu_id(pybind11::object gpu) {
+        gpu_id = py2optional<int>(gpu, "gpu id must be None, or a non-negative integer.", is_nonneg_int);
+    };
+
+    void set_num_threads(int threads) {
+        pyarb::assert_throw(is_nonneg_int(threads), "threads must be a non-negative integer.");
+        num_threads = threads;
+    };
+
+    const opt_int get_gpu_id()      const { return gpu_id; }
+    const int     get_num_threads() const { return num_threads; }
+    const bool    has_gpu()         const { return (gpu_id.value_or(-1) >= 0) ? true : false; }
+
+    // helper function to use arb::make_context(arb::proc_allocation)
+    arb::proc_allocation allocation() const {
+        return arb::proc_allocation(num_threads, gpu_id.value_or(-1));
+    }
+};
+
+// Helper template for printing C++ optional types in Python.
+// Prints either the value, or None if optional value is not set.
+template <typename T>
+std::string to_string(const arb::util::optional<T>& o) {
+    if (!o) return "None";
+
+    std::stringstream s;
+    s << *o;
+    return s.str();
+}
+
+std::string proc_alloc_string(const proc_allocation_shim& a) {
+    std::stringstream s;
+    s << "<hardware resource allocation: threads " << a.num_threads
+      << ", gpu id " << to_string(a.gpu_id);
+    s << ">";
+    return s.str();
+}
+
+// helper functions to wrap arb::make_context
+arb::context make_context_shim(const proc_allocation_shim& Alloc) {
+    return arb::make_context(Alloc.allocation());
+}
+
+template <typename Comm>
+arb::context make_context_shim(const proc_allocation_shim& Alloc, Comm comm) {
+    return arb::make_context(Alloc.allocation(), comm);
+}
+
 void register_contexts(pybind11::module& m) {
     using namespace std::string_literals;
     using namespace pybind11::literals;
+    using opt_int = arb::util::optional<int>;
+#ifdef ARB_MPI_ENABLED
```

Comment:
The `opt_mpi_comm` type can be removed if you use `auto`, see comments below.

Reply:
Done.